drivers/kscan/kscan_ft5336.c

```diff
@@ -35,6 +41,21 @@ LOG_MODULE_REGISTER(ft5336, CONFIG_KSCAN_LOG_LEVEL);
 /* REG_Pn_XH: Position */
 #define POSITION_H_MSK		0x0FU
 
+#ifdef CONFIG_KSCAN_FT5336_INTERRUPT
```

Comment:
I see excessive use of "#ifdef & co" here. Please revise it.

Reply:
I believe I cleaned it up about as much as I could.

Since there are three different cases, I thought having this block at the beginning of the code made sense, to determine early on:
- whether we need a timer or not
- on which IRQ edge(s) we must react

I did have a little extra clutter there, however. Thanks.

Reply:
This is not resolved and I do not understand why you mark it as such. All these ifdef constructs here are not easy to read. I also have no understanding that one makes such an effort with macros just to save a few octets in RAM and flash. I think this driver should have only two options, interrupt mode and polling mode, or better just interrupt mode. And why do you need KSCAN_FT5336_PERIOD in the interrupt mode? Why can not interrupt mode not just always support up and down events?

Reply:
Sorry if I clicked on 'resolve issue' when I shouldn't have. I was waiting for some feedback, and I though that maybe you were waiting for me to click on that in order to signal to you that I have addressed your concern.

I don't believe it's really possible to make this code simpler except by reducing the number of handled cases (which you mention). Currently, there are three cases:
- polling mode
- interrupt mode
- interrupt polling mode

Interrupt mode samples only at the highest possible sampling rate. For that reason, it is undesirable for low duty-cycle applications.

Polling mode has poor performance. But it should probably not be eliminated, because some hardware devices may not have a dedicated interrupt line.

Interrupt polling mode has configurable sampling rate, making it much more appropriate for low duty-cycle applications. I personally believe that interrupt mode is rendered obsolete by interrupt polling mode. But I didn't feel like it was up to me to make this decision and remove it. For this reason, I added the third case (interrupt polling mode), keeping the two others. This mode is by far the best solution for the case where only down and up events are necessary. It reacts immediately, and if the sampling period is set to 0, it almost always uses zero MCU time.

I don't understand why you say I made a massive effort with macros to save a few octets in RAM. I didn't do that at all. I simply added the third case in what I thought was a reasonably readable way, because I didn't feel like it was up to me to delete 'interrupt mode'. Of course, if you wish, I could remove the `KSCAN_FT5336_GPIO_INT_EDGE` and `KSCAN_FT5336_ENABLE_TIMER` defines which are the root cause of your apparent discontent. That would simply move the mess further down into the heart of the code, making easy-to-read code like the following much messier:
```
#ifdef KSCAN_FT5336_ENABLE_TIMER
	/** Timer (polling mode or interrupt polling mode). */
	struct k_timer timer;
#endif
```

Concerning your question about why KSCAN_FT5336_PERIOD exists in interrupt polling mode: it is used to select the sampling period. In 'interrupt mode', this is not configurable, and for hardware-related reasons, the device is always sampled at approximately 20ms.

In interrupt-polling mode, KSCAN_FT5336_PERIOD allows the developer to select an appropriate value to satisfy application requirements. If the developer selects zero, no sampling is done at all between down and up events, so the MCU use is almost always zero. Perhaps you could try out the code and compare the three modes while watching the SPI bus on a logic analyzer. The advantage of interrupt polling mode might become more clear to you.

What do you suggest I do? I personally believe that this code is an improvement over what was previously available. Would you like me to delete 'interrupt mode'? That would result in minor simplifications of the code, and I could then delete the additional #defines mentioned above that seem to disturb you. I don't really see the point. Despite the three cases, the file is reasonably readable. And there may be cases where the developer doesn't care about the MCU hogging that happens with interrupt mode, and wishes to sample at the highest possible sampling rate (I believe this could almost be achieved, however, by using interrupt polling mode with KSCAN_FT5336_PERIOD=1.

@MaureenHelm, @gmarull, do you believe that this driver handles too many cases, resulting in messy code? Should any of the three cases (polling, interrupt, interrupt polling) be removed?